### DISCUSSION TITLE
Disallow deleting a member of a form collection

### DISCUSSION THREAD

**Discussion by YLITC** [jrief/django-formset#113] [General]
When using a formset factory in Django you can use the parameter [can_delete](https://docs.djangoproject.com/en/5.0/topics/forms/formsets/#can-delete) to control whether to allow deleting of the formset members. Can this be done in Django-formset? 
It would be nice to have the ability for each member of the collection control whether it can be deleted or not. 
If this cannot be done currently in Django-formset what would be the best way for me implement this?

**Comment by jrief** [comment]
Interesting use-case.

I somehow have a problem with this from a UX perspective: A user shall have always have the possibility to undo an action which he previously executed. If by accident he clicked onto the "Add …"-button, then he wouldn't have that possibility anymore and this can lead to frustration and even panic.

**Comment by YLITC** [comment]
My use case is a Facility with Units. Every facility when created automatically gets a default unit with a 'special' name. This one should not be deleted OR changed. A User can add or delete Units - certainly the ones they add - but not the default one. Yet I do want the default one to show up in the list of units when viewing the facility.

**Comment by jrief** [comment]
I assume that you have a definition similar to this:

```python

class UnitForm(ModelForm):
    # some fields

class UnitCollection(FormCollection):
    min_siblings = 1
    max_siblings = ...
    unit = UnitForms(…)

class FacilityForm(ModelForm):
    # some fields

class FacilityCollection(FormCollection):
    facility = FacilityForm()
    units = UnitCollections()
```

How about refactoring the `FacilityCollection` into:

```python
class FacilityCollection(FormCollection):
    facility = FacilityForm()
    default_unit = UnitForm()
    units = UnitCollections()
```

you then achieve what you want, without having to change the core implementation of django-formset.

**Reply by YLITC** [reply]
good approximation of what I have :+1: 
I was thinking something along that line. Let me try it. But what about ordering ...

**Comment by jrief** [comment]
Well, ordering than only works for the non-default `UnitForm`. If that faeture is required as well, then yes, I would need to add an extra flag `can_delete` to each form. However, I'm somehow reluctant on this.

**Reply by YLITC** [reply]
What is the reluctance? if I may ask

**Reply by jrief** [reply]
not very convinced (yet) that this is the right approach.

**Comment by YLITC** [comment]
I just can't get it work!
I have something like:
```
class UnitCollection(FormCollection):
    min_siblings = 1
    max_siblings = ...
    legend = "Units"
    unit = UnitForms(…)

class FacilityCollection(FormCollection):
    facility = FacilityForm()
    default_unit = UnitForm()
    units = UnitCollections()
```
Problem is that the default_unit keeps getting the data from the parent instance; the Facility. I have tried setting the instance which didn't work. Will using `model_to_dict` work - though I am not sure how it will work - can you help me? 
Regardless seems like more code and complication then having a `can delete` option!
Also the legend (in UnitCollection) will only be on the units and will not include the default unit.

**Reply by jrief** [reply]
Please add your forms/collections to `/testapp` and point me onto your fork of the project. I then will have a look at it.
